### DISCUSSION TITLE
kiali-operator - kiali-deploy.yml does not properly create supported_kiali_images

### DISCUSSION THREAD

**Discussion by jonuhal** [kiali/kiali#4998] [General]
When trying to use the environment variables with the kiali-operator (v1.49.0) to override the `supported_kiali_images`, the file kiali-operator/playbooks/kiali-deploy.yaml playbook file runs the task "Override the supported images if found in the environment" to create the override hash from `RELATED_IMAGE_kiali_v1_49`, but creates a wrongly formatted `supported_kiali_images` fact.

The file `kiali-operator/playbooks/default-supported-images.yml` has examples of the format, such as:

> v1.48:  {"image_name": "quay.io/kiali/kiali", "image_version": "v1.48"}

But the kiali-deploy.yaml playbook task "Override the supported images if found in the environment" creates the following output:

> "v1.48": {"image_name": "gcr.io/my-project/my-env/kiali/kiali", "image_version": "gcr.io/my-project/my-env/kiali/kiali"}

The "image_version" code is not just pushing the value of `item.key` into the "image_version", but is pushing the "image_name" into both places. [The line causing this can be found here](https://github.com/kiali/kiali-operator/blob/v1.49/playbooks/kiali-deploy.yml#L33).

**Comment by jmazzitelli** [comment]
Can you provide more details about how you are using this RELATED_IMAGE env var feature?

This feature is working fine when used within the context of OpenShift Service Mesh. See: https://github.com/kiali/kiali-operator/blob/master/manifests/kiali-ossm/manifests/kiali.clusterserviceversion.yaml#L274-L281
where ${KIALI_1_48_TAG} is replaced with things like like ":1.2.3"

**Comment by jmazzitelli** [comment]
Without knowing how you are using the feature (what the values are of your env vars, how you are setting them within the operator), the only thing I can think of is you are not formatting the value correctly. You will need to set, say, `RELATED_IMAGE_kiali_default` to a value something like of `gcr.io/my-project/my-env/kiali/kiali:v1.48` where it contains two parts - part 1 is the repo/image name, and part 2 is the version tag - the first part separated from the second part with a colon (`:`).

**Comment by jonuhal** [comment]
I am trying to use Kiali / Kiali Operator within a private GKE cluster, as well as targetting air-gapped clusters. I am planning on having to deploy all container images locally into a container registry and having to reference that local image repo when running workloads from the cluster. I am unable to count on internet access consistently or at all.

I am deploying kiali-operator with a helm chart to include things like istio related resources at the same time. My values.yaml file has a dependency of kiali-operator (v1.49.0) and has several overrides, but specifically for this issue the following:

```
kiali-operator:
  env:
    - name: RELATED_IMAGE_kiali_v1_48
      value: gcr.io/my-project/my-env/kiali/kiali
```

When running with this environment variable, I can see the value set in the environment for the kiali-operator:

```
$ kubectl exec -it -n my-kiali my-kiali-kiali-operator-75f74c686b-mcqbd -- env
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
HOSTNAME=my-kiali-kiali-operator-75f74c686b-mcqbd
TERM=xterm
ANSIBLE_CONFIG=/etc/ansible/ansible.cfg
RELATED_IMAGE_kiali_v1_48=gcr.io/my-project/my-env/kiali/kiali
WATCH_NAMESPACE=
POD_NAME=my-kiali-kiali-operator-75f74c686b-mcqbd
POD_NAMESPACE=my-kiali
ANSIBLE_DEBUG_LOGS=true
ANSIBLE_VERBOSITY_KIALI_KIALI_IO=1
KUBERNETES_SERVICE_HOST=10.216.0.1
KUBERNETES_SERVICE_PORT=443
KUBERNETES_SERVICE_PORT_HTTPS=443
KUBERNETES_PORT=tcp://10.216.0.1:443
KUBERNETES_PORT_443_TCP=tcp://10.216.0.1:443
KUBERNETES_PORT_443_TCP_PROTO=tcp
KUBERNETES_PORT_443_TCP_PORT=443
KUBERNETES_PORT_443_TCP_ADDR=10.216.0.1
container=oci
OPENSHIFT_CI=true
BUILD_LOGLEVEL=0
HOME=/opt/ansible
USER_NAME=ansible
USER_UID=1001
OPENSHIFT_BUILD_NAME=ansible-operator
OPENSHIFT_BUILD_NAMESPACE=ci-op-235bi1sy
```

But when I check the logs, I see the following output for the "Override the supported images if found in the environment" Ansible task:

```
 TASK [Override the supported images if found in the environment] ******************************** 
ok: [localhost] => (item={'key': 'v1.48', 'value': {'image_name': 'quay.io/kiali/kiali', 'image_version': 'v1.48'}}) => {"ansible_facts": {"supported_kiali_images": {"default": {"image_name": "quay.io/kiali/kiali", "image_version": "operator_version"}, "v1.24": {"image_name": "quay.io/kiali/kiali", "image_version": "v1.24"}, "v1.36": {"image_name": "quay.io/kiali/kiali", "image_version": "v1.36"}, "v1.48": {"image_name": "gcr.io/my-project/my-env/kiali/kiali", "image_version": "gcr.io/my-project/my-env/kiali/kiali"}}}, "ansible_loop_var": "item", "changed": false, "item": {"key": "v1.48", "value": {"image_name": "quay.io/kiali/kiali", "image_version": "v1.48"}}}
```

The part `"v1.48": {"image_name": "gcr.io/my-project/my-env/kiali/kiali", "image_version": "gcr.io/my-project/my-env/kiali/kiali"}` is where I see the issue. The "image_version" should be the same as the "key" value of the section and not a repeat of the "image_name" value.

**Comment by jmazzitelli** [comment]
(converting to a discussion - as this is not a bug)

```
    - name: RELATED_IMAGE_kiali_v1_48
      value: gcr.io/my-project/my-env/kiali/kiali
```

And that would be the problem and explains what you are seeing. The related image env var declares the repo name AND the version tag that goes with it. See my previous comment about the format of this value.

**Comment by jmazzitelli** [comment]
And just to be clear, this is not necessarily true: `The "image_version" should be the same as the "key" value of the section `

The key value defines the version (used to match that which the user wants to use via the Kiali CR `spec.version`) but it is technically independent of (does not need to equal) the image version tag. You could have, for example, "v1.48" point to your air-gapped image known as `gcr.io/my-project/my-env/kiali/kiali:v1.48custompatch4` or `gcr.io/my-project/my-env/kiali/kiali:foobar` or `gcr.io/my-project/my-env/kiali/kiali:v1.48.9`

**Comment by jonuhal** [comment]
Okay, then it was just my misunderstanding of how it is supposed to work. I saw the format of the other items coming out of that Ansible task and assumed it should be in the same format, split between "image_name" and "image_version". 

Also, I found in my values.yaml file in my helm chart where I was defining `kiali-operator.cr.spec.deployment.image_version`, which was also causing me issues as well. I had seen the error message in the Kiali-operator log stating:

```
TASK [default/kiali-deploy : Only allow ad-hoc kiali image when appropriate] ***
fatal: [localhost]: FAILED! => {"changed": false, "msg": "The operator is forbidden from accepting a Kiali CR that defines an ad hoc Kiali image [gcr.io/my-project/my-env/kiali/kiali:v1.48.0]. Remove spec.deployment.image_name, spec.deployment.image_version, and spec.deployment.image_digest from the Kiali CR."}
```

... but it didn't click that I had defined this in that file because I was hung up on the "supported_kiali_images" Ansible variable.

By the way, thanks for the help @jmazzitelli. I really appreciate it!

**Reply by jmazzitelli** [reply]
> The operator is forbidden from accepting a Kiali CR that defines an ad hoc Kiali image [gcr.io/my-project/my-env/kiali/kiali:v1.48.0]. 

To allow a user to override the operator default images by specifying his own Kiali server image (aka an "ad-hoc image") in the Kiali CR , you have to tell the operator to allow that (the operator does not allow this by default for security reasons) - essentially you want to tell the operator "I trust my users not to install rogue images into the cluster.

That's what these allow-ad-hoc settings are for: https://github.com/kiali/helm-charts/blob/v1.49.0/kiali-operator/values.yaml#L57-L68

See also:
